vpr/src/route/rr_graph_reader.cpp

```diff
@@ -46,779 +34,754 @@
-void process_nodes(pugi::xml_node parent, const pugiutil::loc_data& loc_data);
-void process_edges(pugi::xml_node parent, const pugiutil::loc_data& loc_data, int *wire_to_rr_ipin_switch, const int num_rr_switches);
-void process_channels(t_chan_width& chan_width, pugi::xml_node parent, const pugiutil::loc_data& loc_data);
+#include <libxml/parser.h>
+
+typedef std::unordered_map<std::string, std::string> AttributeMap;
+
+/********************* Subroutines local to this module *******************/
+void load_rr_file(const char *name);
+void on_start_element(void *ctx, const xmlChar *_name, const xmlChar **_attrs);
+void on_end_element(void *ctx, const xmlChar *name);
+void on_characters(void *ctx, const xmlChar *ch, int len);
+
+void consume_channels(AttributeMap& attrs);
+void consume_switches(AttributeMap& attrs);
+void consume_segments(AttributeMap& attrs);
+void consume_block_types(AttributeMap& attrs);
+void consume_grid(AttributeMap& attrs);
+void consume_rr_nodes(AttributeMap& attrs);
+void consume_rr_edges(AttributeMap& attrs);
+void consume_channel(AttributeMap& attrs);
+void consume_x_list(AttributeMap& attrs);
+void consume_y_list(AttributeMap& attrs);
+void consume_switch(AttributeMap& attrs);
+void consume_switch_timing(AttributeMap& attrs);
+void consume_switch_sizing(AttributeMap& attrs);
+void consume_segment(AttributeMap& attrs);
+void consume_segment_timing(AttributeMap& attrs);
+void consume_block_type(AttributeMap& attrs);
+void consume_pin_class(AttributeMap& attrs);
+void consume_pin(AttributeMap& attrs);
+void consume_grid_loc(AttributeMap& attrs);
+void consume_node(AttributeMap& attrs);
+void consume_node_loc(AttributeMap& attrs);
+void consume_node_timing(AttributeMap& attrs);
+void consume_node_segment(AttributeMap& attrs);
+void consume_node_metadata(AttributeMap& attrs);
+void consume_meta(AttributeMap& attrs);
+void consume_edge(AttributeMap& attrs);
+void consume_edge_metadata(AttributeMap& attrs);
+
+void process_edges(int *wire_to_rr_ipin_switch, int num_rr_switches);
 void process_rr_node_indices(const DeviceGrid& grid);
-void process_seg_id(pugi::xml_node parent, const pugiutil::loc_data & loc_data);
-void set_cost_indices(pugi::xml_node parent, const pugiutil::loc_data& loc_data,
-        const bool is_global_graph, const int num_seg_types);
-
-/************************ Subroutine definitions ****************************/
+void set_cost_indices(bool is_global_graph, const int num_seg_types);
+void process_seg_id();
+
+/********************* Global variables for this module *******************/
+static std::stack <const char *> parser_stack;
+static xmlSAXHandler sax_handler = {};
+
+/* We need to build the current <meta> node piece by piece here. */
+static t_offset current_meta_o = {};
+static std::string current_meta_name;
+static std::string current_meta_value;
+static enum {NODE, EDGE} current_meta_place;
+
+/* Edges create trouble when processed sequentially, so we read all into memory.
```

Comment:
Can you explain what the trouble actually is?

Reply:
It's due to the `_edges` field being a `unique_ptr` in the `t_rr_node` struct. We just cannot push an edge into it. It has to be allocated first and assigned later.

Reply:
Please put the explanation in your code comment

Reply:
Apparently they can be pushed now.